pkgs/by-name/cm/cmake/setup-hook.sh

```diff
@@ -1,5 +1,14 @@
 addCMakeParams() {
-    addToSearchPath CMAKE_PREFIX_PATH $1
+    addToSearchPath CMAKE_INCLUDE_PATH "$1/include"
+    addToSearchPath CMAKE_INCLUDE_PATH "$1"
+
+    addToSearchPath CMAKE_LIBRARY_PATH "$1/lib"
+    addToSearchPath CMAKE_LIBRARY_PATH "$1"
+
+    addToSearchPath CMAKE_PREFIX_PATH "$1/lib/cmake"
+    addToSearchPath CMAKE_PREFIX_PATH "$1/share/cmake"
+    addToSearchPath CMAKE_PREFIX_PATH "$1/lib"
+    addToSearchPath CMAKE_PREFIX_PATH "$1/share"
```

Comment:
There are some packages that put executables in `/lib` and `/share` as a hacky kind of `/libexec`; I’m a little worried that could cause issues? Still a big improvement over the status quo though obviously.

Reply:
We could probably fix `extra-cmake-modules` to install into `/lib/cmake` if that’s the only issue with `/lib`. Did you encounter anything that needed bare `/share`?

Reply:
(FWIW I don’t feel fully confident in the CMake hook stuff enough to get this through to the finish line myself, but if @SomeoneSerge is willing hopefully we can collaborate on getting this merged?)

Reply:
> We could probably fix `extra-cmake-modules` to install into `/lib/cmake` if that’s the only issue with `/lib`. Did you encounter anything that needed bare `/share`?

in the early commit i have annotated that vulkan headers / vulkan-loader needed it. given that these are default search paths for cmake removing them would well, remove the default search paths

> There are some packages that put executables in /lib and /share as a hacky kind of /libexec; I’m a little worried that could cause issues? Still a big improvement over the status quo though obviously.

the change i had which patched cmake should deal with that, i think: https://github.com/NixOS/nixpkgs/pull/318226/commits/52fa78963bf3615fe5e677667a2a021ac628b1f9 it doesn't add any paths from CMAKE_PREFIX_PATH when a binary is requested.  I opted to not use that change as i was triggering the new nixpkgs behaviour with an env var but i thought that would confuse people if they were trying to do builds via nix-shell and invoking cmake by hand and not via runPhase.

Reply:
Thanks for elaborating. I think the patch behaviour seems like it’d get us closer to where we want to be overall. Maybe we could set the `NIXPKGS_CMAKE_DISABLE_BIN_PREFIX` variable in a hook instead of directly in the `cmake` invocation? Nix development shells have a bunch of impurities and weirdnesses anyway, so I think it’s okay if we just make the happy path work.

Reply:
> but if @SomeoneSerge is willing hopefully we can collaborate on getting this merged?

I'm very much interested in getting this merged. I expect a _lot_ of fall out, e.g. I'd expect there to be projects which abuse some kind of generic `find_file` which we'd break, but I'd rather we tightened the bolts and handled those as special cases

Reply:
> > but if @SomeoneSerge is willing hopefully we can collaborate on getting this merged?
> 
> I'm very much interested in getting this merged. I expect a _lot_ of fall out, e.g. I'd expect there to be projects which abuse some kind of generic `find_file` which we'd break, but I'd rather we tightened the bolts and handled those as special cases

i built all attributes that had cmake in the nativeBuildInputs and strict deps set and fixed the breakage
https://github.com/NixOS/nixpkgs/pull/324545
https://github.com/NixOS/nixpkgs/pull/324951

along with darwin up to, i forget how far i built. kdePackages builds 400 packages or something that all set strictDeps but many of them don't work with strictDeps so they need to be disabled. I did all of these builds more than a month ago so probably need to do another round.

Reply:
- added a PR to remove strictDeps from kdePackages https://github.com/NixOS/nixpkgs/pull/329822